python/ray/serve/controller.py

```diff
@@ -435,15 +442,78 @@ def get_root_url(self):
                 )
         return http_config.root_url
 
+    def config_checkpoint_deleted(self) -> bool:
+        """Returns whether the config checkpoint has been deleted.
+
+        Get the config checkpoint from the kv store. If it is None, then it has been
+        deleted.
+        """
+        return self.kv_store.get(CONFIG_CHECKPOINT_KEY) is None
+
     def shutdown(self):
-        """Shuts down the serve instance completely."""
+        """Shuts down the serve instance completely.
+
+        This method will only be triggered when `self._shutting_down` is true. It
+        deletes the kv store for config checkpoint, set application state to deleting,
+        delete all deployments, and shutdown all HTTP proxies. If all the above
+        resources are released, it will then kill the controller actor.
+        """
+        if not self._shutting_down:
+            return
+
+        logger.info("Controller shutdown started!", extra={"log_to_stderr": False})
         self.kv_store.delete(CONFIG_CHECKPOINT_KEY)
         self.application_state_manager.shutdown()
         self.deployment_state_manager.shutdown()
         self.endpoint_state.shutdown()
         if self.http_state:
             self.http_state.shutdown()
 
+        config_checkpoint_deleted = self.config_checkpoint_deleted()
+        application_is_shutdown = self.application_state_manager.is_shutdown()
+        deployment_is_shutdown = self.deployment_state_manager.is_shutdown()
+        endpoint_is_shutdown = self.endpoint_state.is_shutdown()
+        http_state_is_shutdown = (
+            self.http_state is None or self.http_state.is_shutdown()
+        )
+        if (
+            config_checkpoint_deleted
+            and application_is_shutdown
+            and deployment_is_shutdown
+            and endpoint_is_shutdown
+            and http_state_is_shutdown
+        ):
+            logger.warning(
+                "All resources are shutdown, shutting down controller!",
+                extra={"log_to_stderr": False},
+            )
+            _controller_actor = ray.get_actor(
+                self.controller_name, namespace=SERVE_NAMESPACE
+            )
+            ray.kill(_controller_actor, no_restart=True)
```

Comment:
Tried calling `ray.actor.exit_actor()` to kill the controller, but seems like all it does it throwing a special error. Controller continue to exist with or without try-except block. `ray.kill()` seems to be the only way to kill this gracefully, but let me know if this is not ideal. 🙏

Reply:
`ray.kill()` should be fine here.

Reply:
You can use `ray.get_runtime_context().current_actor` instead of fetching it by name.

Reply:
Cool good to know, will change it!